core/src/container.cpp

```diff
@@ -806,46 +814,144 @@ void Alternatives::onNewSolution(const SolutionBase& s) {
 	liftSolution(s);
 }
 
+Fallbacks::Fallbacks(const std::string& name) : Fallbacks(new FallbacksPrivate(this, name)) {}
+
+Fallbacks::Fallbacks(FallbacksPrivate* impl) : ParallelContainerBase(impl) {}
+
 void Fallbacks::reset() {
-	active_child_ = nullptr;
 	ParallelContainerBase::reset();
 }
 
 void Fallbacks::init(const moveit::core::RobotModelConstPtr& robot_model) {
 	ParallelContainerBase::init(robot_model);
-	active_child_ = pimpl()->children().front().get();
 }
 
 bool Fallbacks::canCompute() const {
-	while (active_child_) {
-		StagePrivate* child = active_child_->pimpl();
-		if (child->canCompute())
-			return true;
+	auto impl { pimpl() };
 
-		// active child failed, continue with next
-		auto next = child->it();
-		++next;
-		if (next == pimpl()->children().end())
-			active_child_ = nullptr;
+	if (impl->requiredInterface() == GENERATE) {
+		// current_generator_ is fixed if it produced solutions before
+		if( !solutions().empty() )
+			return (*impl->current_generator_)->pimpl()->canCompute();
 		else
-			active_child_ = next->get();
+			// we still have children to try
+			return impl->current_generator_ != impl->children().end();
 	}
-	return false;
+	else
+		return !pimpl()->pending_states_.empty();
 }
 
 void Fallbacks::compute() {
-	if (!active_child_)
+	auto impl { pimpl() };
+
+	if(impl->requiredInterface() == GENERATE)
+		impl->computeGenerate();
+	else
+		impl->computeFromExternal();
+}
+
+void Fallbacks::onNewSolution(const SolutionBase& s) {
+	liftSolution(s);
+}
+
+FallbacksPrivate::FallbacksPrivate(Fallbacks* me, const std::string& name)
+   : ParallelContainerBasePrivate(me, name) {}
+
+void FallbacksPrivate::initializeExternalInterfaces(InterfaceFlags expected) {
+	if (expected & READS_START)
+		starts().reset(new Interface([this](Interface::iterator external, bool updated) {
+		   this->onNewExternalState<Interface::FORWARD>(external, updated);
+		}));
+	if (expected & READS_END)
+		ends().reset(new Interface([this](Interface::iterator external, bool updated) {
+		   this->onNewExternalState<Interface::BACKWARD>(external, updated);
+		}));
+
+	// we've got to set this somewhere once the interface flags are known.
+	// so we might as well do it here
+	if(expected == GENERATE)
+		current_generator_ = children().begin();
+}
+
+void FallbacksPrivate::onNewFailure(const Stage& child, const InterfaceState* from, const InterfaceState* to) {
+	// only react to failure if it's the last possible candidate failing
+	// otherwise there might still be a feasible solution
+	if(&child == &*children().back())
+		ContainerBasePrivate::onNewFailure(child, from, to);
+}
+
+void FallbacksPrivate::computeGenerate() {
+	if(solutions_.empty())
+		// move to first generator that can run
+		while(current_generator_ != children().end() && !(*current_generator_)->pimpl()->canCompute())
+			++current_generator_;
```

Comment:
Yes, this Fallback container must not be used with `MonitoringGenerator`, the previous implementation had the same limitation.
I plan to have different logic (possible even a `MonitoringFallbacks` container) for this case, because you would expect all Monitors in there to monitor the same Stage and the `Fallbacks` container to pass on solutions individually as well.

I agree about better scheduling required, but I have to tackle one task after the other and `Fallbacks` was much higher up on my list right now and my family life reduces my effective work time even more than usual at the moment...